helix-term/src/ui/menu.rs

```diff
@@ -29,30 +32,36 @@ pub trait Item: Sync + Send + 'static {
         let label: String = self.format(data).cell_text().collect();
         label.into()
     }
+
+    fn tie_breaker(&self, _data: &Self::Data) -> Self::TieBreaker;
```

Comment:
But I don't like having this extra function, it should be unnecessary because this is the intent of `sort_text` (should really be renamed as sort_key).

Let's try modifying the return value of `sort_text` as `K: Ord`, and use a default that uses `Cow<str>`. Then for your one specific implementation you can instead return `(label, language_server_id)` (`(Cow<str>, usize)`)

Reply:
I agree we don't want an extra method. I am not sure what the right API is here though.

A couple points:

* for the picker the concept of sort text does not make sense IMO. In general I think the needs of the picker and of the menu keep diverging (some of mikes colum related work for example) so I think we will likely just endup with two traits (or even something totally different for the picker)
* for the menu I am not sure how much sense all of this actually makes. It feels wrong to me to sort by language server first since fuzzy matching is not very selective so you would see all the "bad" completions from one server before those from a second server. A score bonus is not a good solution because it's hard to implement (how large should the bonuss shpuld be is subjctive and would depend on word length/content) and ultimately not feel predictable. I proposed making it a tie breaker but exactly identicsl fuzzy scores are not as common as you would think fue to the fairly complex bpnus system. It's probably the least bad option right now. 
* all of this doesn't even take into account yet that we do not actually want to sort for a lot of servers. With incomplete completion request support we have to rerequest completions on every keypress and are not supposed to do filtering and scoring ourself. The question then becomes how to mix those servers with other completion options. The server sorts internslly according to their own criteria and scoring system which we are unaware of and also not comparable. This feature has a really large impact and is very commonly used (and the capability is not respected, for example by RA) so I see it as fairly high priority so such an obvious conflict here is a concern for me. 

This makes it hard for me to review since I dont understand the full problem space yet. We may endup needing a very different solution and changing the trait before we have figured that out feels premature. Perhaphs sorting by server first works after all if we filter items with a score below 50% of the top score or something. Ideally we should study what other ides (primarily vscode) do here. There is also a whole can of worms of super long LSP issue trackers where it was discussed how filtering and sorting should work exactly (which can sometimes not have a clear co conclusion,  I need to read trough all that carefully again).

One thing I am fairly sure about is that we want this to be (language_server, label) and not the other way around. Score ties are already rare (which is the only situation where sort_text matters) and when they do occur we want to group by language sever first (otherwise the language sever comparison basically never matters anyway).

Reply:
> One thing I am fairly sure about is that we want this to be (language_server, label)

I was assuming that too, but that probably doesn't make sense either since you could have very weak matches from your priority server that would now appear before better matches from a second one.

Reply:
Ah no that actually would not happen (currently). Sort text is not used for fuzzy matching. It's only used as a tie breaker if the fuzzy match score is equal.


So sorting would be `(fuzzy_score, language_server, label)` so the order of these two only takes effect when the score is the same (which is already pretty rare)